core/src/main/java/com/google/zxing/oned/Code128Writer.java

```diff
@@ -282,66 +298,242 @@ private static CType findCType(CharSequence value, int start) {
+    public int chooseCode(CharSequence value, int start, int oldCode) {
+      CType lookahead = findCType(value, start);
+      if (lookahead == CType.ONE_DIGIT) {
+        if (oldCode == CODE_CODE_A) {
+          return CODE_CODE_A;
+        }
+        return CODE_CODE_B;
+      }
+      if (lookahead == CType.UNCODABLE) {
+        if (start < value.length()) {
+          char c = value.charAt(start);
+          if (c < ' ' || (oldCode == CODE_CODE_A && (c < '`' || (c >= ESCAPE_FNC_1 && c <= ESCAPE_FNC_4)))) {
+            // can continue in code A, encodes ASCII 0 to 95 or FNC1 to FNC4
+            return CODE_CODE_A;
+          }
+        }
+        return CODE_CODE_B; // no choice
+      }
+      if (oldCode == CODE_CODE_A && lookahead == CType.FNC_1) {
         return CODE_CODE_A;
       }
+      if (oldCode == CODE_CODE_C) { // can continue in code C
+        return CODE_CODE_C;
+      }
+      if (oldCode == CODE_CODE_B) {
+        if (lookahead == CType.FNC_1) {
+          return CODE_CODE_B; // can continue in code B
+        }
+        // Seen two consecutive digits, see what follows
+        lookahead = findCType(value, start + 2);
+        if (lookahead == CType.UNCODABLE || lookahead == CType.ONE_DIGIT) {
+          return CODE_CODE_B; // not worth switching now
+        }
+        if (lookahead == CType.FNC_1) { // two digits, then FNC_1...
+          lookahead = findCType(value, start + 3);
+          if (lookahead == CType.TWO_DIGITS) { // then two more digits, switch
+            return CODE_CODE_C;
+          } else {
+            return CODE_CODE_B; // otherwise not worth switching
+          }
+        }
+        // At this point, there are at least 4 consecutive digits.
+        // Look ahead to choose whether to switch now or on the next round.
+        int index = start + 4;
+        while ((lookahead = findCType(value, index)) == CType.TWO_DIGITS) {
+          index += 2;
+        }
+        if (lookahead == CType.ONE_DIGIT) { // odd number of digits, switch later
+          return CODE_CODE_B;
+        }
+        return CODE_CODE_C; // even number of digits, switch now
+      }
+      // Here oldCode == 0, which means we are choosing the initial code
+      if (lookahead == CType.FNC_1) { // ignore FNC_1
+        lookahead = findCType(value, start + 1);
+      }
+      if (lookahead == CType.TWO_DIGITS) { // at least two digits, start in code C
+        return CODE_CODE_C;
+      }
       return CODE_CODE_B;
     }
-    if (lookahead == CType.UNCODABLE) {
-      if (start < value.length()) {
-        char c = value.charAt(start);
-        if (c < ' ' || (oldCode == CODE_CODE_A && (c < '`' || (c >= ESCAPE_FNC_1 && c <= ESCAPE_FNC_4)))) {
-          // can continue in code A, encodes ASCII 0 to 95 or FNC1 to FNC4
-          return CODE_CODE_A;
+  }
+
+  /** 
+   * Encodes minimally using Divide-And-Conquer with Memoization
+   **/
+  private static class MinimalSelector implements Selector {
```

Comment:
Like I'd kind of prefer the separate new code live in its own top level class, for clarity

Reply:
Sure, no problem. This version was the option with the most code reuse in my opinion. By reimplementing only the lookahead function it shared the checks (checks for maximum length and check for illegal characters), the the building of the pattern, the weighted checksum accumulation, and the adding of the checksum and the stop pattern to the patterns.
To make it less invasive I see two alternative options.
1) Initially I wrote a version (also without introducing a new top level class) that has two code paths that compute the "pattern" ArrayList. That way it also shared the checks and the adding of the stop pattern (the checksum adding could also be shared but then the function would need to return two values (the ArrayList and the checksum) .
2) We rename Code128Writer.java to Code128FastWriter.java and add a new class Code128MinimalWriter.java. The checking code could be put in a separate function in Code128FastWriter.java and then be reused by calling it from Code128MinimalWriter too.
I think that both options are fine. The checksum computation and the adding of the stop pattern are so trivial that I don't think that it is worth the bother to try to share them so that I lean slightly more towards option 2. 
Which would you prefer or do you see yet another good way of doing it?

Reply:
Update: On second thought, I am not so fond of option 2 after looking at the code some more. I propose to come up with a solution that leaves the exiting encode() function completely without modification and just dispatches between two versions of the function. This way the two functions can share the constants and the standard overrides of OneDimensionalCodeWriter. I will make a proposal and then we can discuss it from there.

Reply:
Update: I created an new PR #1484 with a less intrusive version.